### PR TITLE
Bug 1435834: Integrate Focus blocking stats code 

### PR DESCRIPTION
**Pull request by garvankeeley**
https://bugzilla.mozilla.org/show_bug.cgi?id=1435834

This is the non-UI changes required for blocking statistics.
Test case included.

### CODE REVIEW

**Comment by garvankeeley** [comment]
In db559c6 I moved the JS injection to an earlier event, but I also moved the night mode injection. I don't think we have any issues where night mode JS isn't injected before the page starts to render, but if so, this would fix that as it now injects before any document JS runs.

**Comment by garvankeeley** [comment]
@justindarc PR updated

File: Client/Frontend/Browser/UserScriptManager.swift [hunk 1]
```diff
@@ -5,33 +5,23 @@
...
-        }
 
-        // Main Frame (.atDocumentStart)
-         if let path = Bundle.main.path(forResource: "MainFrameAtDocumentStart", ofType: "js"),
-             let source = try? NSString(contentsOfFile: path, encoding: String.Encoding.utf8.rawValue) as String {
-             let userScript = WKUserScript(source: source, injectionTime: .atDocumentStart, forMainFrameOnly: true)
-             tab.webView?.configuration.userContentController.addUserScript(userScript)
-         }
+    // Scripts can use this to verify the app –not js on the page– is calling into them.
+    public static let securityToken = UUID()
```
Comment (garvankeeley) [inline Client/Frontend/Browser/UserScriptManager.swift]: @boek this token gets injected into every script.

File: Client/Frontend/UserContent/UserScripts/MainFrame/AtDocumentStart/TrackingProtectionStats.js [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+    writable: false,
+    value: { enabled: false }
+  });
+
+  Object.defineProperty(window.__firefox__.TrackingProtectionStats, "setEnabled", {
+    enumerable: false,
+    configurable: false,
+    writable: false,
+    value: function(enabled, securityToken) {
+      if (securityToken !== SECURITY_TOKEN) {
```
Comment (garvankeeley) [inline Client/Frontend/UserContent/UserScripts/MainFrame/AtDocumentStart/TrackingProtectionStats.js]: @boek this func requires the security token

File: Client/Frontend/Browser/UserScriptManager.swift [hunk 3]
```diff
@@ -5,33 +5,23 @@
...
+    init(tab: Tab) {
+        [(WKUserScriptInjectionTime.atDocumentStart, mainFrameOnly: false),
+         (WKUserScriptInjectionTime.atDocumentEnd, mainFrameOnly: false),
+         (WKUserScriptInjectionTime.atDocumentStart, mainFrameOnly: true),
+         (WKUserScriptInjectionTime.atDocumentEnd, mainFrameOnly: true)].forEach { arg in
+            let (position, mainFrameOnly) = arg
+            let name = (mainFrameOnly ? "MainFrame" : "AllFrames") + "AtDocument" + (position == .atDocumentStart ? "Start" : "End")
+            if let path = Bundle.main.path(forResource: name, ofType: "js"),
+               let source = try? NSString(contentsOfFile: path, encoding: String.Encoding.utf8.rawValue) as String {
+                let wrappedSource = "(function() { var SECURITY_TOKEN = '\(UserScriptManager.securityToken)'; \(source) })()"
```
Comment (garvankeeley) [inline Client/Frontend/Browser/UserScriptManager.swift]: SECURITY_TOKEN injection in the closure scope of the IIFE

File: Client/Frontend/Browser/TabManager.swift [hunk 4]
```diff
@@ -828,18 +828,26 @@ extension TabManager {
...
-        tab?.setNightMode(isNightMode)
+        tab.setNightMode(isNightMode)
 
         if #available(iOS 11, *) {
             let isNoImageMode = self.prefs.boolForKey(PrefsKeys.KeyNoImageModeStatus) ?? false
-            tab?.noImageMode = isNoImageMode
+            tab.noImageMode = isNoImageMode
+
+            if let tpHelper = tab.contentBlocker as? ContentBlockerHelper, !tpHelper.isEnabledForTab {
+                webView.evaluateJavaScript("window.__firefox__.TrackingProtectionStats.setEnabled(false, \(UserScriptManager.securityToken))", completionHandler: nil)
```
Comment (garvankeeley) [inline Client/Frontend/Browser/TabManager.swift]: Disable stats JS at first opportunity.

File: Client/Frontend/UserContent/UserScripts/MainFrame/AtDocumentStart/TrackingProtectionStats.js [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+    Array.prototype.map.call(document.images, function(t) { return t.src }).forEach(sendMessage);
+  }
+
+  let originalOpen = null;
+  let originalSend = null;
+  let originalImageSrc = null;  
+  let mutationObserver = null;
+
+  function injectStatsTracking(enabled) {
+    if (enabled) {
```
Comment (garvankeeley) [inline Client/Frontend/UserContent/UserScripts/MainFrame/AtDocumentStart/TrackingProtectionStats.js]: Changes from focus: I added the enable/disable block, and moved the postload js injection into a window.addEventListener(load)

File: Client/Frontend/UserContent/UserScripts/MainFrame/AtDocumentStart/TrackingProtectionStats.js [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+        })
+        originalImageSrc.set.call(this, value);
+      }
+    });
+
+    // -------------------------------------------------
+    // Listen to when new <script> elements get added to the dom
+    // and send the source to the host application
+    // -------------------------------------------------
+    mutationObserver = new MutationObserver(function(mutations) {
```
Comment (justindarc) [inline Client/Frontend/UserContent/UserScripts/MainFrame/AtDocumentStart/TrackingProtectionStats.js]: A couple things here:
1. I'd like to see if we can just hijack `HTMLScriptElement.prototype.src` like we do for `Image` above?
2. Does hijacking `Image.prototype.src` above actually work for appending ``s to the DOM in all cases? i.e.: `var img = document.createElement('img'); img.src = '/foo.jpg'; document.body.appendChild(img);`
3. If we do actually have to use a `MutationObserver` for `<script>` elements, we need to at least check for the existence of `src` before calling `postMessage()`. For instance, I think it's possible to append a `<script>` to the DOM that has inline JS.

File: Client/Frontend/ContentBlocker/ContentBlockerHelper.swift [hunk 7]
```diff
@@ -421,3 +432,30 @@ extension ContentBlockerHelper {
...
 }
+
+@available(iOS 11, *)
+extension ContentBlockerHelper : TabContentScript {
+    class func name() -> String {
+        return "TrackingProtectionStats"
+    }
+
+    func scriptMessageHandlerName() -> String? {
+        // The JS is used verbatim from Focus, so use their handler name.
```
[the remark this answers is not in the document]
Comment (garvankeeley) [reply Client/Frontend/ContentBlocker/ContentBlockerHelper.swift]: Yeah, not copy-pasted anymore, might as well change this message name and remove the comment, will do

File: Client/Frontend/UserContent/UserScripts/MainFrame/AtDocumentStart/TrackingProtectionStats.js [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+      }
+
+      window.__firefox__.TrackingProtectionStats.enabled = enabled;
+
+      injectStatsTracking(enabled);
+    }
+  })
+
+  function onLoadNativeCallback() {
+    var messageHandler = window.webkit.messageHandlers.focusTrackingProtection;
```
[the remark this answers is not in the document]
Comment (garvankeeley) [reply Client/Frontend/UserContent/UserScripts/MainFrame/AtDocumentStart/TrackingProtectionStats.js]: Will do

File: Client/Frontend/UserContent/UserScripts/MainFrame/AtDocumentStart/TrackingProtectionStats.js (hunk 6 above)
Comment (justindarc) [reply Client/Frontend/UserContent/UserScripts/MainFrame/AtDocumentStart/TrackingProtectionStats.js]: You can test those cases with this JSBin: http://output.jsbin.com/rahizoq

**Pull request merged**